dataplane/src/pipeline/pipeline.rs

```diff
@@ -44,15 +108,54 @@ impl<Buf: PacketBufferMut + 'static> DynPipeline<Buf> {
     /// [`nf_dyn`]
     #[allow(unused)]
     pub fn add_stage_dyn(mut self, nf: Box<dyn DynNetworkFunction<Buf>>) -> Self {
-        self.nfs.push(nf);
+        self.internal_add_stage_dyn_with_id(AutoStageId::Auto(self.nfs.len()), nf);
         self
     }
+
+    pub fn add_stage_dyn_with_id(
+        &mut self,
+        id: Id,
+        nf: Box<dyn DynNetworkFunction<Buf>>,
+    ) -> Result<&mut Self, PipelineError> {
+        self.internal_add_stage_dyn_with_id(AutoStageId::Id(id), nf)
+    }
+
+    fn internal_add_stage_dyn_with_id(
+        &mut self,
+        id: AutoStageId<Id>,
+        nf: Box<dyn DynNetworkFunction<Buf>>,
+    ) -> Result<&mut Self, PipelineError> {
+        // FIXME(mvachhar): There seems to be no method to insert and error if the key already exists.
```

Comment:
You may be able to achieve something similar with the entry() method and struct Entry.

Reply:
Maybe.  I'm working up a playground to ask on the Rust Discord.  I have most of a solution with `std::borrow` but that requires that the return type of the `borrow` function on the trait returns a reference to `T` in `Borrow<T>`, which causes an issue in trying ot return an `Option` or `Enum` constructed in the function.

Reply:
Unfortunately not.  See [this open issue](https://github.com/rust-lang/rust/issues/82766)